proposals/0184-block-writeable-account-data-limit.md

```diff
@@ -0,0 +1,107 @@
+---
+simd: '0184'
+title: Block Writeable Account Data Limit
+authors:
+  - Brennan Watt (Anza)
+category: Standard
+type: Core
+status: Idea
+created: 2024-10-15
+feature: (fill in with feature tracking issues once accepted)
+supersedes:
+superseded-by:
+extends:
+---
+
+## Summary
+
+A protocol limit on the amount of account data in bytes that may be written
+during a single block.
+
+## Motivation
+
+When direct mapping is enabled, clients will be able to handle much more account
+data because there will be fewer copies in memory (e.g. for each VM invocation /
+CPI call). However, writeable account data cannot be direct mapped because it
+needs to be written back after a successful tx execution. This necessitates a
+decoupling between account read and write data in order to increase readable
+account data limits.
+
+Note: There is an existing per block limit on newly allocated account data
+(100MB), but this does not encompass account data being updated/written back.
+
+## New Terminology
+
+None
+
+## Detailed Design
+
+This feature would check the amount of account data marked as writeable for each
+transaction being handled by the block producer (for inclusion in a block) or
+validator (replaying transactions in a block), add it to a running total of
+total writeable account data for the block, and compare it against some
+threshold (recommend 2GB to start) to decide if the transaction is allowed to be
+included in the block (if sum is below the limit for the case of block producer)
+or if the slot needs to be marked dead (if sum is above the limit for the case
+of validator).
```

Comment:
andrew will have to look at this because if it can invalidate blocks then it may interfere with asynchronous execution

Reply:
its unfortunately quite different, transaction data size (sum of all accounts required for a transaction) can be calculated on the transaction level, and if the transaction exceed its limit, it is treated as invalid. so in a world with async execution, the transaction could be included in a block before verifying it, but all nodes will see that it violates its limits and treat it the same (charge fees and not execute it, once the feature gate for fee-only transactions is active)

but if we added a data size limit at the block level, you would need to know the running total of all previous transactions to determine whether you can put the transaction in the block. the purpose of removing constraints to enable async execution is that block-packers which put "bad" transactions in a block might make suboptimal blocks, but they can always be sure the block is valid if they follow constraints that dont depend on execution (are transactions syntactically valid, are blockhashes valid, etc)

i believe we could have a block-level constraint as proposed here if instead of "if you exceed the block data size limit, the block is invalid" it was like "if you exceed the block data size limit, no further transactions in the block will be executed." this could be user-hostile tho, because it lets block producers invalidate transactions that should otherwise succeed. especially if we charge fees, the block producer now has an incentive to exceed the limit and farm fees without having to execute transactions

alternatively we could very easily charge more for writable data at the transaction level, but im not sure that accomplishes the goals of this simd. andrew may have better suggestions than me tho

Reply:
yep, think we are on the same page! the way transaction data size works now is theres a default budget (64MiB), adjustable by compute budget instruction, and at execution time the accounts are totalled and the transaction is invalid if it exceeds its limit. but this doesnt affect the block, so it doesnt interfere with async execution

we could add a transaction-level writable accounts budget with a smaller limit (assume 2MiB like you suggest), and it would likewise be async-safe. we could easily have block-level constraints based on those requested budgets, which would also be async-safe. the main risk is people requesting enormous budgets and crowding other transactions out of the block, which we could defray with either a per-transaction maximum limit** or by exponentially ramping the cost per byte

if you make changes to the transaction data size calculation code, note that the formula is quite tricky. i have an open simd to fix this #186 

**(16-20MiB would allow writing to a 10Mib account plus plenty of breathing room, but i dont actually know how big the largest accounts are onchain. 10MiB is the max allocation size but i believe you can pump accounts larger than that with realloc)